src/binder/statement/create_table.rs

```
@@ -45,57 +45,73 @@ impl Binder {
+                        // primary key is declared only by "primary key(c1, c2) syntax"
+                        let pos = *(ordered_pk_pos.get(col.name()).unwrap());
+                        ordered_pk_ids[pos] = idx as ColumnId;
+                        col.set_primary(true); // TODO: remove this line in the future
                     }
+                    col.set_id(idx as ColumnId);
+                    col_catalogs.push(col);
                 }
 
-                let columns = columns
-                    .iter()
-                    .enumerate()
-                    .map(|(idx, col)| {
-                        let mut col = ColumnCatalog::from(col);
-                        if extra_pk_name.contains(col.name()) && !col.is_primary() {
-                            col.set_primary(true);
-                        }
-                        col.set_id(idx as ColumnId);
-                        col
-                    })
-                    .collect();
                 Ok(BoundCreateTable {
                     database_id: db.id(),
                     schema_id: schema.id(),
                     table_name: table_name.into(),
-                    columns,
+                    columns: col_catalogs,
+                    ordered_pk_ids,
                 })
             }
             _ => panic!("mismatched statement type"),
         }
     }
+
+    fn get_extra_pks(constraints: &Vec<TableConstraint>) -> HashMap<String, usize> {
+        let mut ordered_pk_pos: HashMap<String, usize> = HashMap::new();
+        for constraint in constraints {
+            match constraint {
+                TableConstraint::Unique {
+                    is_primary,
+                    columns,
+                    ..
+                } if *is_primary => columns.iter().enumerate().for_each(|(index, indent)| {
+                    ordered_pk_pos.insert(indent.value.clone(), index);
+                }),
+                _ => todo!(),
```

Better to return an error, so that the program won't panic.

Also, will there be multiple `unique` constraints in parser AST? We should only allow single unique appearing in a statement, and return error otherwise.

---

Also I think the signature of this function is somehow confusing... The `HashMap<String, usize>` usize part seems like position of column in unique constraint. But generally, developers want a column id at this stage. For example,

```
create table (v1 int, v2 int, primary key(v2));
```

... where currently we will generate v2 -> 0. But what we need is something like `pk: vec![column id #1]`.

I think we can simply return Vec<Ident> for this function, and let the other part of the code to handle the mapping between ident and column id.

---

Also I think the signature of this function is confusing... The `HashMap<String, usize>` usize part seems like position of column in unique constraint. But generally, developers want a column id at this stage. For example,

```
create table (v1 int, v2 int, primary key(v2));
```

... where currently we will generate v2 -> 0. But what we need is something like `pk: vec![column id #1]`.

I think we can simply return Vec<Ident> for this function, and let the other part of the code to handle the mapping between ident and column id.